### ISSUE TITLE
Use less mutation

### ISSUE THREAD

**Issue by antoine-levitt** [JuliaMolSim/DFTK.jl#50]
Eg in `update_hamiltonian!`. See discussion in https://github.com/mfherbst/DFTK.jl/pull/45#issuecomment-549728257

**Comment by mfherbst** [comment]
Key point: Make hamiltonian fully immutable, i.e. change both potential and density arrays *at each update* with a new density. In brief: Cut `update_hamiltonian!` completely for `update_hamiltonian`.

**Comment by antoine-levitt** [comment]
Possibly not-so-radical proposal: cut off mutation completely from the code, except in the utilities and the eigensolver. Getting rid of that would go a long towards enabling us to use Zygote, which would be very nice.

I'm not very concerned about memory churn. Essentially, anything that doesn't involve orbitals is just noise, either in terms of performance or memory usage, so we can do what we want there. 

Grepping for "!" is illuminating: we don't actually use that much mutation (although we do have "hidden" (non-!) mutation in a couple of places, most notably the terms). Possibly the most annoying place is energies, but I'm not that happy with the current interface anyway.

**Comment by antoine-levitt** [comment]
Reopening this and pinging @niklasschmitz for input. As far as I can tell, most of the places where we use mutation, we don't actually need it and could replace with more functional patterns. The exceptions are for performance-sensitive functions, where we use eg `.+=` quite a bit. Should we try and get rid of mutation as far as we can?

**Comment by antoine-levitt** [comment]
Possibly problematic patterns, outside of the clearly mutating functions (eg lowpass_for_symmetries! and friends):

```
        occupation[ik] = zeros(T, n_bands)
        occupation[ik][1:n_fill] .= filled_occ
```
```
for ...
cond && push!(res)
end
```
(with possibly multiple pushes per iteration)
```
a = zeros(...)
for i in ...
a[i] = ...
```
(not always easy to replace by higher-level functions, see eg PspHgh line 105)
```
E = zeros(T)
for k in kpoints
E += ...
end
```
with each iteration independent

**Comment by antoine-levitt** [comment]
Also noticed when playing around with this: one needs to be careful when reformulating in a functional style, as that can really push the limits of inference, and needs to be checked with `@code_warntype` (for the low-level functions that are performance-sensitive)

**Comment by mfherbst** [comment]
> Should we try and get rid of mutation as far as we can?

Yes we should, but to be fair quite a few of the patterns you list above we use quite a lot and they lead to rather nice to read and modular code. I think about the XC for example where there are quite a few branches for LDA, GGA, possibly multiple functionals looped over etc. So if at all possible we should only change this if it is necessary and not impacting to much the readability.

BTW: I think I would call this hijacking an old issue, because the original concern has almost nothing to do with what we discuss here. I'm not a fan of this, but ok for this time.

**Comment by antoine-levitt** [comment]
Yeah, we should discuss what kinds of pattern need to be changed and whether we compromise on readability.
